lite2/client.go

```diff
@@ -707,46 +625,64 @@ func (c *Client) sequence(
 }
 
 // see VerifyHeader
+// Bisection finds the middle header between a trusted and new header, reiterating the action until it
+// verifies a header. A cache of headers requested by the primary is kept such that when a
+// verification is made, and the light client tries again to verify the new header,
+// the light client does not need to ask for all the same headers again.
 func (c *Client) bisection(
 	initiallyTrustedHeader *types.SignedHeader,
 	initiallyTrustedVals *types.ValidatorSet,
 	newHeader *types.SignedHeader,
 	newVals *types.ValidatorSet,
 	now time.Time) error {
 
+	type headerSet struct {
+		sh     *types.SignedHeader
+		valSet *types.ValidatorSet
+	}
+
 	var (
+		headerCache = []headerSet{{newHeader, newVals}}
+		depth       = 0
+
 		trustedHeader = initiallyTrustedHeader
 		trustedVals   = initiallyTrustedVals
-
-		interimHeader = newHeader
-		interimVals   = newVals
 	)
 
 	for {
 		c.logger.Debug("Verify newHeader against trustedHeader",
 			"trustedHeight", trustedHeader.Height,
 			"trustedHash", hash2str(trustedHeader.Hash()),
-			"newHeight", interimHeader.Height,
-			"newHash", hash2str(interimHeader.Hash()))
+			"newHeight", headerCache[depth].sh.Height,
+			"newHash", hash2str(headerCache[depth].sh.Hash()))
 
-		err := Verify(c.chainID, trustedHeader, trustedVals, interimHeader, interimVals, c.trustingPeriod, now,
-			c.trustLevel)
+		err := Verify(c.chainID, trustedHeader, trustedVals, headerCache[depth].sh, headerCache[depth].valSet,
+			c.trustingPeriod, now, c.trustLevel)
 		switch err.(type) {
 		case nil:
-			if interimHeader.Height == newHeader.Height {
+			// Have we verified the last header
+			if depth == 0 {
 				return nil
 			}
-
-			// Update the lower bound to the previous upper bound
-			trustedHeader, trustedVals = interimHeader, interimVals
-			// Update the upper bound to the untrustedHeader
-			interimHeader, interimVals = newHeader, newVals
+			// If not, update the lower bound to the previous upper bound
+			trustedHeader, trustedVals = headerCache[depth].sh, headerCache[depth].valSet
+			// Remove the untrusted header at the lower bound in the header cache - it's no longer useful
+			headerCache = headerCache[:depth]
+			// Reset the cache depth so that we start from the upper bound again
+			depth = 0
 
 		case ErrNewValSetCantBeTrusted:
-			pivotHeight := (interimHeader.Height + trustedHeader.Height) / 2
-			interimHeader, interimVals, err = c.fetchHeaderAndValsAtHeight(pivotHeight)
-			if err != nil {
-				return err
+			// do add another header to the end of the cache
+			if depth == len(headerCache)-1 {
+				pivotHeight := (headerCache[depth].sh.Height + trustedHeader.Height) / 2
+				interimHeader, interimVals, err := c.fetchHeaderAndValsAtHeight(pivotHeight)
+				if err != nil {
+					return err
+				}
+				depth++
+				headerCache = append(headerCache, headerSet{interimHeader, interimVals})
+			} else { // or we shift to the next header in the cache
```

Comment:
can't we execute depth++ after if and drop else branch? we have it in both branches.

Reply:
Yup we can do that